encoding/graph/type_/vertex_generator.rs

```diff
@@ -4,27 +4,85 @@
  * file, You can obtain one at https://mozilla.org/MPL/2.0/.
  */
 
-use std::sync::atomic::{AtomicU16, Ordering};
+use std::sync::atomic::AtomicU16;
+use std::sync::atomic::Ordering::Relaxed;
+use bytes::Bytes;
+use primitive::prefix_range::PrefixRange;
+use resource::constants::snapshot::BUFFER_KEY_INLINE;
 
-use storage::{snapshot::WriteSnapshot, MVCCStorage};
+use storage::key_value::{StorageKey, StorageKeyReference};
 use storage::snapshot::WritableSnapshot;
 
 use crate::{
     graph::type_::vertex::{
-        build_vertex_attribute_type, build_vertex_attribute_type_prefix, build_vertex_entity_type,
-        build_vertex_entity_type_prefix, build_vertex_relation_type, build_vertex_relation_type_prefix,
-        build_vertex_role_type, build_vertex_role_type_prefix, TypeID, TypeVertex,
+        build_vertex_attribute_type, build_vertex_entity_type, build_vertex_relation_type,
+        build_vertex_role_type, TypeID, TypeVertex,
     },
     Keyable,
 };
+use crate::error::EncodingError;
+use crate::error::EncodingErrorKind::{ExhaustedTypeIDs, FailedTypeIDAllocation};
+use crate::graph::type_::Kind;
+use crate::graph::type_::Kind::{Attribute, Entity, Relation, Role};
+use crate::graph::type_::vertex::TypeIDUInt;
+use crate::graph::Typed;
+
+pub struct TypeIDAllocator {
+    root_type : Kind,
+    last_allocated_type_id: AtomicU16,
+    to_vertex:  fn (TypeID) -> TypeVertex<'static>
```

Comment:
The fn could be replaced by the struct reference above if needed.
The root type could also be stored in the struct instead.